torch/csrc/distributed/spmd/event_impl.cpp

```diff
@@ -0,0 +1,33 @@
+#include <torch/csrc/distributed/spmd/event_impl.h>
+
+#include <torch/types.h>
+
+namespace torch {
+namespace distributed {
+namespace spmd {
+
+namespace {
+
+// NB: need to call torch::class_ to register event classes in the map returned
+// by c10::getCustomClassTypeMap(). Otherwise, these classes cannot be wrapped
+// by an IValue.
+// NB: add this line here instead of in spmd/init.cpp because these classes are
+// not yet meant to be visible from Python.
+static const auto prepareModuleEvent =
+    torch::class_<PrepareModuleEvent>("spmd", "_PrepareModuleEvent");
+static const auto preForwardEvent =
+    torch::class_<PreForwardEvent>("spmd", "_PreForwardEvent");
+static const auto localgradReadyEvent =
+    torch::class_<LocalGradReadyEvent>("spmd", "_LocalGradReadyEvent");
+static const auto bucketReadyEvent =
+    torch::class_<BucketReadyEvent>("spmd", "_BucketReadyEvent");
+static const auto commDoneEvent =
+    torch::class_<CommDoneEvent>("spmd", "_CommDoneEvent");
+static const auto globalGradReadyEvent =
+    torch::class_<GlobalGradReadyEvent>("spmd", "_GlobalGradReadyEvent");
```

Comment:
Why do we need to create these torch custom classes? Aren't all of the event handlers actually executing in cpp?

Reply:
that's a limitation in IValue custom type registration. `torch::class_` currently is the only way to register the string name of custom type. We can also modify JIT code in future PRs to expose a dedicated name registration API.